### PR TITLE
Tim Sort

### PR DESCRIPTION
**Pull request by dqian96**
Implementation of Tim Sort for the Sort module - WIP

Currently implemented:
- Binary insertion sort (including correctness and performance)

Next steps:
- Tim sort without galloping

Benchmarking:

```
Time taken to sort 65536 bytes (8192 int(64)s)
quickSort (seconds): 0.022325
heapSort (seconds): 0.027805
insertionSort (seconds): 1.78996
binaryInsertionSort (seconds): 0.89685
mergeSort (seconds): 0.097196
selectionSort (seconds): 1.97712
```

### CODE REVIEW

**Comment by ben-albrecht** [comment]
Thanks for the contribution @dqian96!

We'll get a reviewer on this sometime soon hopefully.

In the meantime, be sure to run `make docs` and confirm that the Sort documentation (`doc/html/modules/packages/Sort.html`) renders as you expect, if you haven't already.

Also, it's generally good practice to run [`chplspell`](https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/SpellChecking.rst) to catch spelling errors in comments (which propagate to documentation in modules). I noticed 1 spelling error at a quick glance, so running this script might save both you and the reviewer some time. 

Before we're able to merge your contribution, we'll need you to submit a [CLA](https://github.com/chapel-lang/chapel/tree/master/doc/rst/developer/contributorAgreements/). Let me know if you have any questions.

File: modules/packages/Sort.chpl [hunk 1]
```diff
@@ -511,6 +511,64 @@ proc insertionSort(Data: [?Dom] ?eltType, comparator:?rec=defaultComparator)
...
+  for i in low..high by stride {
+    var ithVal = Data[i];
+
+    var lo: int = low,
+        hi: int = i - stride,
+        loc: int = -1;              // index that ithVal should be inserted at
+
+    while (lo <= hi) {
+      const size = (hi - lo) / stride,
+            mid = if hi == lo then hi
```
Comment (daviditen) [inline modules/packages/Sort.chpl]: I think this can be simplified to be just `mid = lo + (size/2)*stride`.

**Comment by dqian96** [comment]
@ben-albrecht Thanks for pointing me to those useful tools!

@daviditen Good catches on the binary search and graphs. I implemented those changes on the last commit.

I decided to factor out binary search for the last occurrence of an element into a private proc since Tim Sort will use it extensively.

I'll continue with implementing Tim Sort in the next week or so. Thanks!

**Comment by dqian96** [comment]
Thanks for the review and merge!!

I will create a new branch to work on the remainder of Tim Sort this weekend, which uses binary insertion sort heavily. In that branch, I will fix the 80 char limit and trailing whitespace.

**Pull request merged**